### PR TITLE
Revive PR: Allow healthchecking multiple RabbitMQ connections

### PR DESCRIPTION
**Pull request by cieciurm**
**What this PR does / why we need it**:
This PR revives #1659

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #764

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public IConnection? Connection { get; set; }
+
+    /// <summary>
+    /// A connection factory for RabbitMQ.
+    /// </summary>
+    public IConnectionFactory? ConnectionFactory { get; set; }
+
+    /// <summary>
+    /// An Uri representing a  connection string for RabbitMQ.
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs:21]:
```suggestion
    /// An Uri representing a connection string for RabbitMQ.
```

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    /// <summary>
+    /// A connection factory for RabbitMQ.
+    /// </summary>
+    public IConnectionFactory? ConnectionFactory { get; set; }
+
+    /// <summary>
+    /// An Uri representing a connection string for RabbitMQ.
+    /// </summary>
+    /// <remarks>
+    /// Can be used in conjunction with the <see cref="SslOption"/> property.
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs:24]:
```suggestion
    /// Can be used in conjunction with the <see cref="Ssl"/> property.
```

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheck.cs [hunk 3]
```diff
@@ -6,31 +7,22 @@ namespace HealthChecks.RabbitMQ;
...
+    private IConnection? _connection;
+    private readonly RabbitMQHealthCheckOptions _options;
 
-    public RabbitMQHealthCheck(IConnectionFactory factory)
+    public RabbitMQHealthCheck(RabbitMQHealthCheckOptions options)
     {
-        _factory = Guard.ThrowIfNull(factory);
-        _ownsConnection = true;
-    }
+        _options = options;
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheck.cs:19]:
```suggestion
        _options = Guard.ThrowIfNull(options);
```

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheck.cs [hunk 4]
```diff
@@ -49,47 +41,29 @@ public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, Canc
...
+                        AutomaticRecoveryEnabled = true,
+                        UseBackgroundThreadsForIO = true,
+                    };
 
-                if (_sslOption != null)
-                    ((ConnectionFactory)_factory).Ssl = _sslOption;
-            }
+                    if (_options.Ssl is not null)
+                    {
+                        ((ConnectionFactory)factory).Ssl = _options.Ssl;
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheck.cs:60]: Cast. Well, OK. Let keep it as is.

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace HealthChecks.RabbitMQ;
+
+/// <summary>
+/// Options for <see cref="RabbitMQHealthCheck"/>.
+/// </summary>
+public class RabbitMQHealthCheckOptions
+{
+    /// <summary>
+    /// An already created connection to RabbitMQ.
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs:11]:
```suggestion
    /// An already created connection to RabbitMQ.
    /// It has priority over <see cref="ConnectionFactory"/> and <see cref="ConnectionUri"/>.
```

File: src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    /// It has priority over <see cref="ConnectionFactory"/> and <see cref="ConnectionUri"/>.
+    /// </summary>
+    public IConnection? Connection { get; set; }
+
+    /// <summary>
+    /// A connection factory for RabbitMQ.
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/RabbitMQHealthCheckOptions.cs:17]:
```suggestion
    /// A connection factory for RabbitMQ.
    /// It has priority over <see cref="ConnectionUri"/>.
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 7]
```diff
@@ -153,6 +157,10 @@ public static class RabbitMQHealthCheckBuilderExtensions
     /// <summary>
     /// Add a health check for RabbitMQ services using <see cref="IConnectionFactory"/> factory function.
     /// </summary>
+    /// <remarks>
+    /// <paramref name="connectionFactoryFactory"/> will be called each time the healthcheck route is requested. However
+    /// the created <see cref="IConnection"/> will be reused.
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:162]: >      /// the created <see cref="IConnection"/> will be reused.

not true, options object is created each time so `GetOrAdd` falls into delegate creting connection

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 8]
```diff
@@ -185,6 +190,10 @@ public static class RabbitMQHealthCheckBuilderExtensions
     /// <summary>
     /// Add a health check for RabbitMQ services using connection string (amqp uri) factory.
     /// </summary>
+    /// <remarks>
+    /// <paramref name="connectionStringFactory"/> will be called each time the healthcheck route is requested. However
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:194]: the same here

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 9]
```diff
@@ -219,6 +225,10 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:229]: and here

File: test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs [hunk 10]
```diff
@@ -238,4 +238,38 @@ public async Task be_healthy_if_rabbitmq_is_available_using_iServiceProvider()
...
+                app.UseHealthChecks("/health2", new HealthCheckOptions
+                {
+                    Predicate = r => r.Name.Equals("rabbitmq2")
+                });
+            });
+
+        using var server = new TestServer(webHostBuilder);
+
+        var response1 = await server.CreateRequest("/health1").GetAsync().ConfigureAwait(false);
+        var response2 = await server.CreateRequest("/health2").GetAsync().ConfigureAwait(false);
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs:270]:
```suggestion
        using var response1 = await server.CreateRequest("/health1").GetAsync().ConfigureAwait(false);
        using var response2 = await server.CreateRequest("/health2").GetAsync().ConfigureAwait(false);
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs (hunk 7 above)
Comment (sungam3r) [reply src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:162]: To use approach with options as a dictionary key you should move options creation in all methods out of  HealthCheckRegistration's delegate.

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 11]
```diff
@@ -139,12 +152,9 @@ public static class RabbitMQHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        builder.Services
-            .AddSingleton(sp => new RabbitMQHealthCheck(connectionFactory(sp)));
-
         return builder.Add(new HealthCheckRegistration(
             name ?? NAME,
-            sp => sp.GetRequiredService<RabbitMQHealthCheck>(),
+            sp => new RabbitMQHealthCheck(new RabbitMQHealthCheckOptions { Connection = connectionFactory(sp) }),
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:157]: Memory leak here.

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 12]
```diff
@@ -171,12 +185,9 @@ public static class RabbitMQHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        builder.Services
-            .AddSingleton(sp => new RabbitMQHealthCheck(connectionFactoryFactory(sp)));
-
         return builder.Add(new HealthCheckRegistration(
             name ?? NAME,
-            sp => sp.GetRequiredService<RabbitMQHealthCheck>(),
+            sp => new RabbitMQHealthCheck(new RabbitMQHealthCheckOptions { ConnectionFactory = connectionFactoryFactory(sp) }),
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:190]: Memory leak here.

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 13]
```diff
@@ -205,12 +220,9 @@ public static class RabbitMQHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        builder.Services
-            .AddSingleton(sp => new RabbitMQHealthCheck(connectionStringFactory(sp), sslOption));
-
         return builder.Add(new HealthCheckRegistration(
             name ?? NAME,
-            sp => sp.GetRequiredService<RabbitMQHealthCheck>(),
+            sp => new RabbitMQHealthCheck(new RabbitMQHealthCheckOptions { ConnectionUri = connectionStringFactory(sp), Ssl = sslOption }),
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:225]: Memory leak here.

**Comment by cieciurm** [comment]
@sungam3r I gave it some thought and here is my suggestion:
- Have two registration methods for `Options` object.
  * First without external args (pure)
  * Second with `IServiceProvider` dependency
- Now we don't create memory leaks and we allow configuring the `Options` using `IServiceProvider`

Please let me know, what do you think now 😄

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 14]
```diff
@@ -60,12 +60,15 @@ public static class RabbitMQHealthCheckBuilderExtensions
         IEnumerable<string>? tags = default,
         TimeSpan? timeout = default)
     {
-        builder.Services
-            .AddSingleton(sp => new RabbitMQHealthCheck(rabbitConnectionString, sslOption));
+        var options = new RabbitMQHealthCheckOptions
```
[the remark this answers is not in the document]
Comment (cieciurm) [reply src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:63]: Solution: unify registration method, and have two for the whole `Options` object

**Comment by sungam3r** [comment]
I'm fine either way - breaking or not, with 2 overloads or 10.

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs (hunk 12 above)
Comment (sungam3r) [reply src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs]: correct

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 15]
```diff
@@ -165,80 +180,19 @@ public static class RabbitMQHealthCheckBuilderExtensions
     /// <returns>The specified <paramref name="builder"/>.</returns>
     public static IHealthChecksBuilder AddRabbitMQ(
         this IHealthChecksBuilder builder,
-        Func<IServiceProvider, IConnectionFactory> connectionFactoryFactory,
+        Func<IServiceProvider, RabbitMQHealthCheckOptions>? setup,
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:183]: Let's change `Func<IServiceProvider, RabbitMQHealthCheckOptions>` to `Action<IServiceProvider, RabbitMQHealthCheckOptions>`.

File: test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs [hunk 16]
```diff
@@ -98,7 +98,7 @@ public async Task be_healthy_if_rabbitmq_is_available_using_iconnectionfactory()
             {
                 services
                     .AddHealthChecks()
-                    .AddRabbitMQ(sp => factory, tags: new string[] { "rabbitmq" });
+                    .AddRabbitMQ(setup => setup.ConnectionFactory = factory, tags: new string[] { "rabbitmq" });
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs:101]:
```suggestion
                    .AddRabbitMQ(options => options.ConnectionFactory = factory, tags: new string[] { "rabbitmq" });
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs (hunk 15 above)
Comment (cieciurm) [reply src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs]: Thanks, it's been changed

**Comment by cieciurm** [comment]
Thanks, it's been updated accordingly 😄 

cc @sungam3r

**Review by sungam3r — APPROVED** [review]

**Pull request merged**